### PR TITLE
add reasonable toString implementations to domain objects

### LINKED ISSUE #386: Add reasonable toString implementations to domain objects
Having e.g. `com.gooddata.md.Metric@60d8c9b7` in outputs is not very useful.

Consider:
- text representation of fields
- JSON string
- ...

### PR DESCRIPTION
**Pull request by Tantalor93**
resolves #386

### CODE REVIEW

File: src/main/java/com/gooddata/util/GoodDataToStringBuilder.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        setExcludeFieldNames(LINKS_FIELD_NAME);
+    }
+
+    /**
+     * replaces implementation of {@link ReflectionToStringBuilder#toStringExclude(Object, Collection)},
+     * adds default exclusion of links field
+     *
+     * @param object
+     * @param excludeFieldNames
+     * @return
```
Comment (jimirocks) [inline src/main/java/com/gooddata/util/GoodDataToStringBuilder.java]: - [ ] add return description

File: src/main/java/com/gooddata/util/GoodDataToStringBuilder.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                                                  .toString();
+    }
+
+    /**
+     * replaces implementation of {@link ReflectionToStringBuilder#toStringExclude(Object, String...)},
+     * adds default exclusion of links field
+     *
+     * @param object
+     * @param excludeFieldNames
+     * @return
```
Comment (jimirocks) [inline src/main/java/com/gooddata/util/GoodDataToStringBuilder.java]: - [ ] add return description

File: src/main/java/com/gooddata/util/GoodDataToStringBuilder.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import org.apache.commons.lang3.builder.RecursiveToStringStyle;
+import org.apache.commons.lang3.builder.ReflectionToStringBuilder;
+
+import java.util.Collection;
+
+/**
+ * GoodDataToStringBuilder enables to create string representation based on fields of an object and exclude specified fields
+ * links field is excluded by default.
+ */
+public class GoodDataToStringBuilder extends ReflectionToStringBuilder {
```
Comment (jimirocks) [inline src/main/java/com/gooddata/util/GoodDataToStringBuilder.java]: I wonder whether to override also plain `toString(Object)` method

File: src/test/java/com/gooddata/connector/StatusTest.java [hunk 4]
```diff
@@ -66,4 +69,11 @@ public void testIsFailedOnUnknownCode() throws Exception {
         assertThat(new Status("unknown code", "", "").isFailed(), is(false));
     }
 
+    @Test
+    public void testToStringFormat() {
+        final Status status = new Status("unknown code", "", "");
+        final Pattern pattern = Pattern.compile(Status.class.getSimpleName() + "\\[.*\\]");
+        assertThat(pattern.matcher(status.toString()).matches(), CoreMatchers.is(true));
```
Comment (jimirocks) [inline src/test/java/com/gooddata/connector/StatusTest.java]: consider to use `MatchesPattern` matcher

File: src/test/java/com/gooddata/dataload/processes/DataloadProcessesTest.java [hunk 5]
```diff
@@ -6,22 +6,33 @@
...
     public void testDeserialization() throws Exception {
-        final DataloadProcesses processes = new ObjectMapper()
+        processes = new ObjectMapper()
                 .readValue(getClass().getResourceAsStream("/dataload/processes/processes.json"), DataloadProcesses.class);
 
         assertThat(processes, is(notNullValue()));
         assertThat(processes.getItems(), hasSize(1));
     }
 
+    @Test(dependsOnMethods = "testDeserialization")
```
Comment (jimirocks) [inline src/test/java/com/gooddata/dataload/processes/DataloadProcessesTest.java]: I would rather avoid such dependencies in unit tests, @martiner what is your opinion?

File: src/main/java/com/gooddata/util/GoodDataToStringBuilder.java (hunk 3 above)
Comment (Tantalor93) [inline src/main/java/com/gooddata/util/GoodDataToStringBuilder.java]: created new static method `defaultToString(Object, String...)`, which will exclude "links" field by default. Removed override of `toStringExclude(Object, String...)` and `toString(Object, Collection<String>)`.

File: src/test/java/com/gooddata/connector/StatusTest.java (hunk 4 above)
Comment (Tantalor93) [inline src/test/java/com/gooddata/connector/StatusTest.java]: done

File: src/test/java/com/gooddata/dataload/processes/DataloadProcessesTest.java (hunk 5 above)
Comment (Tantalor93) [inline src/test/java/com/gooddata/dataload/processes/DataloadProcessesTest.java]: well, some of these domain objects are only created from deserialization and have no public constructor (most of them has private constructor). I rather have this test dependency, then turn all @JsonCreator constructors visibility to package private and using them manually from tests.

Comment (jimirocks) [inline src/test/java/com/gooddata/dataload/processes/DataloadProcessesTest.java]: to sum up the options (even the absur ones):
1. inter test dependencies
2. make constructors package private
3. do the same deserialization second time for the toString test
4. assert toString in deserialization test

Which of those is the lesser evil? I personally incline to 3., but I am strongly personally biased against test dependencies...

Comment (Tantalor93) [inline src/test/java/com/gooddata/dataload/processes/DataloadProcessesTest.java]: ok, I'll go along with option 3.

**Pull request merged**